src/backend/distributed/commands/alter_table.c

```diff
@@ -1572,3 +1577,67 @@ ExecuteQueryViaSPI(char *query, int SPIOK)
 		ereport(ERROR, (errmsg("could not finish SPI connection")));
 	}
 }
+
+
+/*
+ * SwitchToSequentialAndLocalExecutionIfRelationNameTooLong generates the longest shard name
+ * on the shards of a distributed table, and if exceeds the limit switches to sequential and
+ * local execution to prevent self-deadlocks.
+ *
+ * In case of a RENAME, the relation name parameter should store the new table name, so
+ * that the function can generate shard names of the renamed relations
+ */
+void
+SwitchToSequentialAndLocalExecutionIfRelationNameTooLong(Oid relationId,
+														 char *relationName)
+{
+	if (!IsCitusTable(relationId))
+	{
+		return;
+	}
+
+	char *longestShardName = NULL;
+	if (PartitionedTable(relationId))
+	{
+		/* we may not have any partitions yet, and the function will return NULL in that case */
+		longestShardName = LongestPartitionName(relationId);
+	}
+
+	if (longestShardName == NULL)
+	{
+		longestShardName = pstrdup(relationName);
+	}
+
+	ShardInterval *shardInterval = LoadShardIntervalWithLongestShardName(relationId);
```

Comment:
If this is a partitioned table, should we use the `relationId` of the child table? I guess this part might be broken for `GenerateLongestShardPartitionIndexName` as well.


I guess `LongestPartitionName` should probably be changed to return the `longestPartitionRelationId` and the callers should use that?

Reply:
I think it was clearer to name the function `PartitionWithLongestNameRelationId`. I went with this name